07PerformanceProgramming/Sec04ConcurrencyProblems.md

```diff
@@ -0,0 +1,103 @@
+
+## Race condition
+Another problem that may occur in concurrent programs is that of race conditions. These occur when one thread checks if a resource is available, is interrupted by another that makes that resource unavailable, resumes and is unable to complete the intended action. The effect of this can be anything from starving one of the workers (as above) to causing the entire program or system to crash.
+
+In the case of our philosophers, P3, wants to eat. They check that chopsticks C3 and C4 are available, and picks up C3. In the meantime, P4 has grabbed C4. What happens when P3 tries to pick up the chopstick that is no longer there? In the case of the philosophers, there would no doubt be a full and frank exchange of views. 
+
+### Waiter/arbitrator
+One solution that race conditions is to use an arbitrator for access to the contended resources. In the case of the dining philosophers, this can be thought of as a waiter whom the philosophers ask permission to pick up both their chopsticks. One one is allowed to pick up chopsticks at a time, and the waiter will not allow another to attempt to do so until the previous philosopher has their chopsticks firmly in hand.
+
+The waiter can also resolve the issue of resource starvation by ensuring that not philosopher is allowed to go too long without getting to pick up chopsticks and eat.
+
+In a real computer system, the arbitrator will do much the same thing, controlling access by the worker threads to ensure that access is performed in an orderely fashion, and potentially preventing resource starvation.
+
+In the case of the computer model of the dining philosophers, the arbitrator can be implemented as a mutex.
+
+## Mutex
+A mutex is an object that is shared between threads which allows only one thread to pass a certain point at any one time. The name is short for *mut*ual *ex*clusion, and might also be known as a lock.
+
+A mutex protects access to what is known as a critical section to only one worker thread at a time. In the case of the dining philosophers, this critical section is the action of picking up chopsticks. The worker thread attempts to access the critical section. If no other worker is within, the mutex will allow the worker thread to proceed, perform the actions of the critical section and crucially to release the mutex lock when it is finished. This allows any other thread that might be waiting to proceed into the critical section.
+
+### `std::mutex`
+Introduced in C++11.
+
+The `std::mutex` class provides a minimal, portable interface to the underlying operating system mutexes. It can be included in your code using the `#include <mutex>` standard header inclusion. The base C++ version provides the following member functions:
+* `mutex( )`
+ * default constructor
```

Comment:
Check indentation? On the preview all indents at same level but maybe that doesn't properly preview it

Reply:
Checked and corrected